File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/net/UrlQuery.kt
// Opted in for `PercentEncoding`, which is `@InternalFrakApi`. Per file, not module-wide, so the
// marker still applies elsewhere.
@file:OptIn(InternalFrakApi::class)

package id.frak.sdk.net

import id.frak.sdk.InternalFrakApi
import id.frak.sdk.core.Hex

/**
 * Minimal query-string editing over a URL string. Not `android.net.Uri` (throws on the JVM
 * unit-test classpath) and not `java.net.URI` (re-encodes what is already encoded). Existing
 * parameters are never re-encoded, so links a merchant has already published are unchanged.
 */
internal class UrlQuery private constructor(
    private val base: String,
    private val fragment: String,
    private val parameters: MutableList<Pair<String, String>>,
) {
    /**
     * The value is percent-decoded, and the key match falls back to case-insensitive because
     * channels mangle casing — but an exact match wins, so `?fctx=stale&fCtx=real` resolves to
     * `real`, as `sdk/core/src/utils/url/queryParams.ts` does.
     */
    fun get(key: String): String? =
        (
            parameters.firstOrNull { it.first == key }
                ?: parameters.firstOrNull { it.first.equals(key, ignoreCase = true) }
        )?.second
            ?.let(::percentDecode)

    /**
     * Exact-key lookup, for parameters the web client reads through `URLSearchParams.get`. `fCtx`
     * tolerates mangled casing; `fmt` authorises an identity merge, so it matches web or not at all.
     */
    fun getExact(key: String): String? = parameters.firstOrNull { it.first == key }?.second?.let(::percentDecode)

    fun remove(key: String): UrlQuery =
        apply {
            parameters.removeAll { it.first.equals(key, ignoreCase = true) }
        }

    /** Gap-fill: a merchant's own value always wins. */
    fun fillIfAbsent(
        key: String,
        value: String?,
    ): UrlQuery =
        apply {
            if (value.isNullOrEmpty()) return@apply
            if (get(key) != null) return@apply
            parameters += key to PercentEncoding.encode(value)
        }

    fun set(
        key: String,
        value: String,
    ): UrlQuery =
        apply {
            remove(key)
            parameters += key to PercentEncoding.encode(value)
        }

    override fun toString(): String =
        buildString {
            append(base)
            parameters.forEachIndexed { index, (key, value) ->
                append(if (index == 0) '?' else '&')
                append(key)
                if (value.isNotEmpty()) append('=').append(value)
            }
            append(fragment)
        }

    companion object {
        private const val PERCENT_BYTE: Byte = '%'.code.toByte()
        private const val PLUS_BYTE: Byte = '+'.code.toByte()

        /** Null when [url] has no scheme separator — anything else is treated as an opaque base. */
        fun parse(url: String): UrlQuery? {
            if (!url.contains("://")) return null

            val fragmentAt = url.indexOf('#')
            val fragment = if (fragmentAt >= 0) url.substring(fragmentAt) else ""
            val withoutFragment = if (fragmentAt >= 0) url.substring(0, fragmentAt) else url

            val queryAt = withoutFragment.indexOf('?')
            val base = if (queryAt >= 0) withoutFragment.substring(0, queryAt) else withoutFragment
            val query = if (queryAt >= 0) withoutFragment.substring(queryAt + 1) else ""

            val parameters =
                query
                    .split('&')
                    .filter { it.isNotEmpty() }
                    .map {
                        val separator = it.indexOf('=')
                        if (separator < 0) it to "" else it.substring(0, separator) to it.substring(separator + 1)
                    }.toMutableList()

            return UrlQuery(base, fragment, parameters)
        }

        /**
         * Tolerant by design: a malformed escape is left as written rather than dropping the value.
         * Decodes over UTF-8 bytes, so a non-ASCII character alongside an escape survives verbatim,
         * and `+` decodes to a space as `URLSearchParams` does on the web.
         */
        fun percentDecode(value: String): String {
            if ('%' !in value && '+' !in value) return value
            val bytes = value.toByteArray(Charsets.UTF_8)
            val out = java.io.ByteArrayOutputStream(bytes.size)
            var index = 0
            while (index < bytes.size) {
                val byte = bytes[index]
                val decoded =
                    if (byte == PERCENT_BYTE && index + 2 < bytes.size) {
                        hexByte(bytes[index + 1], bytes[index + 2])
                    } else {
                        null
                    }
                when {
                    decoded != null -> {
                        out.write(decoded)
                        index += 3
                    }

                    byte == PLUS_BYTE -> {
                        out.write(' '.code)
                        index++
                    }

                    else -> {
                        out.write(byte.toInt())
                        index++
                    }
                }
            }
            return out.toString(Charsets.UTF_8.name())
        }

        /** Null unless both bytes are ASCII hex digits: a `toIntOrNull(16)` here accepts a sign, so `%-1` decoded to `0xFF`. */
        private fun hexByte(
            high: Byte,
            low: Byte,
        ): Int? {
            val highNibble = Hex.nibble(high.toInt().toChar()) ?: return null
            val lowNibble = Hex.nibble(low.toInt().toChar()) ?: return null
            return (highNibble shl 4) or lowNibble
        }
    }
}

File: sdk/ios/Sources/FrakSDK/Net/URLQuery.swift
import Foundation

/// Minimal query-string editing over a URL string. Not `URLComponents`, which cannot edit one
/// parameter without re-serialising the whole query through an encoder that re-encodes what is
/// already encoded — and a merchant's published links must not change under them. One exception,
/// which no server distinguishes: an empty value loses its `=`, so `?a=&b=1` round-trips as
/// `?a&b=1`.
struct URLQuery {
    private let base: String
    private let fragment: String
    private var parameters: [(key: String, value: String)]

    /// Nil when `url` has no scheme separator — anything else is treated as an opaque base.
    static func parse(_ url: String) -> URLQuery? {
        guard url.contains("://") else { return nil }

        let withoutFragment: Substring
        let fragment: Substring
        if let hash = url.firstIndex(of: "#") {
            withoutFragment = url[url.startIndex..<hash]
            fragment = url[hash...]
        } else {
            withoutFragment = url[...]
            fragment = ""
        }

        let base: Substring
        let query: Substring
        if let mark = withoutFragment.firstIndex(of: "?") {
            base = withoutFragment[withoutFragment.startIndex..<mark]
            query = withoutFragment[withoutFragment.index(after: mark)...]
        } else {
            base = withoutFragment
            query = ""
        }

        let parameters = query.split(separator: "&").map { part -> (key: String, value: String) in
            guard let separator = part.firstIndex(of: "=") else { return (String(part), "") }
            return (String(part[part.startIndex..<separator]), String(part[part.index(after: separator)...]))
        }

        return URLQuery(base: String(base), fragment: String(fragment), parameters: parameters)
    }

    /// The decoded value at `key`, or nil. Case-folded because channels lowercase query keys in
    /// transit, but an exact match wins, so `?fctx=stale&fCtx=real` resolves to `real`. The value
    /// is percent-decoded because a channel that re-encodes a link turns `-` into `%2D` and the
    /// base64url payload would then fail to decode. Mirrors `utils/url/queryParams.ts`.
    func value(for key: String) -> String? {
        let match =
            parameters.first { $0.key == key }
            ?? parameters.first { $0.key.caseInsensitiveCompare(key) == .orderedSame }
        return match.map { Self.percentDecode($0.value) }
    }

    /// Exact-key lookup, for parameters the web client reads through `URLSearchParams.get`. `fCtx`
    /// tolerates mangled casing; `fmt` authorises an identity merge, so it matches web or not at all.
    func exactValue(for key: String) -> String? {
        parameters.first { $0.key == key }.map { Self.percentDecode($0.value) }
    }

    mutating func remove(_ key: String) {
        parameters.removeAll { $0.key.caseInsensitiveCompare(key) == .orderedSame }
    }

    /// Appends `key` only when it is absent — gap-fill, so a merchant's own value always wins.
    mutating func fillIfAbsent(_ key: String, _ value: String?) {
        guard let value, !value.isEmpty, self.value(for: key) == nil else { return }
        parameters.append((key, PercentEncoding.encode(value)))
    }

    /// Replaces `key`, appending it so ordering stays position-independent.
    mutating func set(_ key: String, to value: String) {
        remove(key)
        parameters.append((key, PercentEncoding.encode(value)))
    }

    var string: String {
        var out = base
        for (index, parameter) in parameters.enumerated() {
            out += index == 0 ? "?" : "&"
            out += parameter.key
            if !parameter.value.isEmpty {
                out += "=" + parameter.value
            }
        }
        return out + fragment
    }

    /// Tolerant by design: a malformed escape is left as written rather than dropping the value.
    /// `+` decodes to a space, as `URLSearchParams` does, so a merchant's own link reads the same
    /// here as it does on the web.
    static func percentDecode(_ value: String) -> String {
        guard value.contains("%") || value.contains("+") else { return value }
        let characters = Array(value.utf8)
        var bytes: [UInt8] = []
        bytes.reserveCapacity(characters.count)
        var index = 0
        while index < characters.count {
            if characters[index] == UInt8(ascii: "%"), index + 2 < characters.count,
                let high = Hex.nibble(characters[index + 1]), let low = Hex.nibble(characters[index + 2])
            {
                bytes.append(high << 4 | low)
                index += 3
            } else if characters[index] == UInt8(ascii: "+") {
                bytes.append(UInt8(ascii: " "))
                index += 1
            } else {
                bytes.append(characters[index])
                index += 1
            }
        }
        return String(decoding: bytes, as: UTF8.self)
    }
}
